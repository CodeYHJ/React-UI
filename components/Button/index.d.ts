import React, { MouseEventHandler, HtmlHTMLAttributes, ReactNode } from 'react';
import "./style/index.less";
declare type ButtonType = 'primary' | 'danger' | 'warn' | 'default' | 'success' | 'dashed';
export interface ButtonProps extends HtmlHTMLAttributes<HTMLButtonElement> {
    type?: ButtonType;
    className?: string;
    onClick?: MouseEventHandler;
    disabled?: boolean;
    pre?: ReactNode;
    loading?: boolean;
}
declare const Button: React.SFC<ButtonProps>;
export default Button;
