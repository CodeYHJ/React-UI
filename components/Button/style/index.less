@import '../../style/var.less';
.buttonBase() {
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  padding: 5px 15px;
  border-radius: 4px;
  border: 1px solid @neutral-color-border;
  cursor: pointer;
  position: relative;
  border: 1px solid @neutral-color-border;
  outline: none;
  background: #ffffff;
}
.hover() {
  &:hover {
    &::before {
      content: '';
      width: 100%;
      height: 100%;
      position: absolute;
      top: 0;
      left: 0;
      background: @neutral-color-background;
    }
  }
}
.code-ui-button {
  height: 30px;
  &-primary {
    .buttonBase();
    background: @main-color-primary;
    color: #ffffff;
    border-color: @main-color-primary;
    .hover();
  }

  &-danger {
    .buttonBase();
    background: @fun-color-danger;
    color: #ffffff;
    border-color: @fun-color-danger;
    .hover();
  }
  &-warn {
    .buttonBase();
    background: @fun-color-warn;
    color: #ffffff;
    border-color: @fun-color-warn;
    .hover();
  }
  &-success {
    .buttonBase();
    background: @fun-color-success;
    color: #ffffff;
    border-color: @fun-color-success;
    .hover();
  }
  &-dashed {
    .buttonBase();
    color: @neutral-color-content;
    border-style: dashed;
    &:hover {
      color: @main-color-primary;
      border-color: @main-color-primary;
    }
    &:focus {
      color: @main-color-primary;
      border-color: @main-color-primary;
    }
  }
  &-default {
    .buttonBase();
    color: @neutral-color-content;
    &:hover {
      color: @main-color-primary;
      border-color: @main-color-primary;
    }
    &:focus {
      color: @main-color-primary;
      border-color: @main-color-primary;
    }
  }
  &-disabled {
    background: @neutral-color-background;
    color: @neutral-color-disabled;
    border: 1px solid @neutral-color-border;
    cursor: not-allowed;

    &:hover {
      color: @neutral-color-disabled;
      border: 1px solid @neutral-color-border;
      &::before {
        content: none;
      }
    }
    &:focus {
      color: @neutral-color-disabled;
      border: 1px solid @neutral-color-border;
    }
  }
  &-loading {
    color: #ffffff;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin-right: 8px;
    animation: button-loading-spin 1s linear infinite;
  }
  &-preIcon {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    margin-right: 8px;
  }
  &-dashed-disabled {
    background: @neutral-color-background;
    color: @neutral-color-disabled;
    border-style: dashed;
    cursor: not-allowed;
    &:hover {
      color: @neutral-color-disabled;
      border: 1px solid @neutral-color-border;
      border-style: dashed;
    }
    &:focus {
      color: @neutral-color-disabled;
      border: 1px solid @neutral-color-border;
      border-style: dashed;
    }
  }
}

@keyframes button-loading-spin {
  0% {
    transform: rotateZ(0);
  }
  100% {
    transform: rotateZ(360deg);
  }
}
