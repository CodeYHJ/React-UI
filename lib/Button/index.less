.buttonBase() {
  display: inline-block;
  padding: 5px 15px;
  border-radius: 4px;
  border: 1px solid grey;
  cursor: pointer;
  position: relative;
}
.hover() {
  &:hover {
    &::before {
      content: "";
      width: 100%;
      height: 100%;
      position: absolute;
      top: 0;
      left: 0;
      background: rgba(255, 255, 255, 0.3);
    }
  }
}
.code-ui-button {
  &-main {
    .buttonBase();
    background: #1890ff;
    color: #ffffff;
    border-color: #1890ff;
    .hover();
  }

  &-danger {
    .buttonBase();
    background: #e8635e;
    color: #ffffff;
    border-color: #e8635e;
    .hover();
  }
  &-warn {
    .buttonBase();
    background: #ebc053;
    color: #ffffff;
    border-color: #ebc053;
    .hover();
  }
  &-safe {
    .buttonBase();
    background: #009900;
    color: #ffffff;
    border: 1px solid #009900;
    border-color: #009900;
    .hover();
  }
  &-default {
    .buttonBase();
    color: grey;
    &:hover {
      color: #1890ff;
      border-color: #1890ff;
    }
  }
}
