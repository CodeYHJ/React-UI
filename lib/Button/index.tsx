import React, { MouseEventHandler, HtmlHTMLAttributes, ReactNode } from 'react';
import "./index.less"
import { classPre } from '@lib/utils';
import { Icon } from '@lib/Icon';

type ButtonType = 'primary' | 'danger' | 'warn' | 'default' | 'success' | 'dashed'
export interface ButtonProps extends HtmlHTMLAttributes<HTMLButtonElement> {
    type?: ButtonType,
    className?: string,
    onClick?: MouseEventHandler,
    disabled?: boolean,
    pre?: ReactNode,
    loading?: boolean
}
const c = classPre('button')

const Button: React.SFC<ButtonProps> = (props) => {
    const { type = 'default', className, disabled, onClick, pre, loading, ...other } = props
    const defaultCls = [c(),c(type), className]
    if (disabled) {
        if (type === 'dashed') {
            defaultCls.push(c('dashed-disabled'))
        } else {
            defaultCls.push(c('disabled'))

        }
    }
    const cls = defaultCls.filter(Boolean).join(' ')
    const handleClick: MouseEventHandler = (e) => {
        if (!disabled) {
            onClick && onClick(e)
        }
    }
    const loadingSvg = () => {
        return (
            <div className={c('loading')}>
                <Icon name="loading"></Icon>
            </div>
        )
    }
    const renderIcon = () => {
        if(disabled)return null
        if (loading) {
            return loadingSvg()
        } else {
            return pre ? (
                <div className={c('preIcon')}>
                    {pre}
                </div>
            ) : null
        }
    }
    return (
        <button className={cls} onClick={handleClick} {...other}>
            {renderIcon()}
            <span>{props.children}</span>
        </button>
    );
}

export default Button;