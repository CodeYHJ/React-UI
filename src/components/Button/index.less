.buttonBase() {
  display: inline-block;
  padding: 5px 15px;
  border-radius: 4px;
  border: 1px solid grey;
  cursor: pointer;
  position: relative;
}
.hover() {
  &:hover {
    &::before {
      content: "";
      width: 100%;
      height: 100%;
      position: absolute;
      top: 0;
      left: 0;
      background: rgba(255, 255, 255, 0.3);
    }
  }
}
.code-ui-button {
  &-main {
    .buttonBase();
    background: #1890ff;
    color: #ffffff;
    border-color: #1890ff;
    .hover();
  }

  &-danger {
    .buttonBase();
    background: #e8635e;
    color: #ffffff;
    border-color: #e8635e;
    .hover();
  }
  &-warn {
    .buttonBase();
    background: #ebc053;
    color: #ffffff;
    border-color: #ebc053;
    .hover();
  }
  &-safe {
    .buttonBase();
    background: #53eb67;
    color: #ffffff;
    border: 1px solid #53eb67;
    border-color: #53eb67;
    .hover();
  }
  &-default {
    .buttonBase();
    color: grey;
    &:hover {
      color: #1890ff;
      border-color: #1890ff;
    }
  }
}
