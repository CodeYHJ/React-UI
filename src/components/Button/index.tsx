import React, { MouseEventHandler } from 'react';
import { classPre } from '@com/utils';
import "./index.less"

type ButtonType = 'main' | 'danger' | 'warn' | 'default' | 'safe'
export interface ButtonProps {
    type?: ButtonType,
    className?: string,
    onClick?: MouseEventHandler
}
const c = classPre('button')

const Button: React.SFC<ButtonProps> = (props) => {
    const { type = 'default', className, ...other } = props
    const cls = [c(type), className].filter(Boolean).join(' ')
    return (
        <div className={cls} {...other}>
            <span>{props.children}</span>
        </div>
    );
}

export default Button;