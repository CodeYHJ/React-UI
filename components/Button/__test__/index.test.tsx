import { mount, shallow } from 'enzyme';

import renderer from 'react-test-renderer';

import React from 'react';

import Button from '../index';

describe('<Button/> Component', () => {
  describe('render', () => {
    it('renders an `.code-ui-button-default` component', () => {
      const json = renderer.create(<Button />).toJSON();

      const wrapper = shallow(<Button />);

      expect(wrapper.find('.code-ui-button-default')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an `.code-ui-button-primary` component', () => {
      const json = renderer.create(<Button type="primary" />).toJSON();

      const wrapper = shallow(<Button type="primary" />);

      expect(wrapper.find('.code-ui-button-primary')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an `.code-ui-button-danger` component', () => {
      const json = renderer.create(<Button type="danger" />).toJSON();

      const wrapper = shallow(<Button type="danger" />);

      expect(wrapper.find('.code-ui-button-danger')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an `.code-ui-button-warn` component', () => {
      const json = renderer.create(<Button type="warn" />).toJSON();

      const wrapper = shallow(<Button type="warn" />);

      expect(wrapper.find('.code-ui-button-warn')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an `.code-ui-button-success` component', () => {
      const json = renderer.create(<Button type="success" />).toJSON();

      const wrapper = shallow(<Button type="success" />);

      expect(wrapper.find('.code-ui-button-success')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an `.code-ui-button-dashed` component', () => {
      const json = renderer.create(<Button type="dashed" />).toJSON();

      const wrapper = shallow(<Button type="dashed" />);

      expect(wrapper.find('.code-ui-button-dashed')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an disabled of dashed', () => {
      const json = renderer.create(<Button type="dashed" disabled />).toJSON();

      const wrapper = shallow(<Button type="dashed" disabled />);

      expect(wrapper.find('.code-ui-button-dashed-disabled')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an disabled of default', () => {
      const json = renderer.create(<Button disabled />).toJSON();

      const wrapper = shallow(<Button disabled />);

      expect(wrapper.find('.code-ui-button-disabled')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders an disabled of primary', () => {
      const json = renderer.create(<Button type="primary" disabled />).toJSON();

      const wrapper = shallow(<Button type="primary" disabled />);

      expect(wrapper.find('.code-ui-button-disabled')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders loading', () => {
      const json = renderer.create(<Button type="primary" loading />).toJSON();

      const wrapper = shallow(<Button type="primary" loading />);

      expect(wrapper.find('.code-ui-button-loading')).toHaveLength(1);

      expect(json).toMatchSnapshot();
    });

    it('renders children when passed in', () => {
      const wrapper = shallow(<Button>测试</Button>);

      expect(wrapper.contains(<span>测试</span>)).toEqual(true);
    });
  });

  describe('event', () => {
    it('should handle onClick Event', () => {
      const onClick = jest.fn();

      const wrapper = mount(<Button onClick={onClick} />);

      wrapper.find('.code-ui-button-default').simulate('click');

      expect(onClick).toBeCalled();
    });
  });
});
