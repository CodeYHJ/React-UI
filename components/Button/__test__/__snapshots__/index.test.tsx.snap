// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> Component render renders an \`.code-ui-button-danger\` component 1`] = `
<button
  className="code-ui-button code-ui-button-danger"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an \`.code-ui-button-dashed\` component 1`] = `
<button
  className="code-ui-button code-ui-button-dashed"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an \`.code-ui-button-default\` component 1`] = `
<button
  className="code-ui-button code-ui-button-default"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an \`.code-ui-button-primary\` component 1`] = `
<button
  className="code-ui-button code-ui-button-primary"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an \`.code-ui-button-success\` component 1`] = `
<button
  className="code-ui-button code-ui-button-success"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an \`.code-ui-button-warn\` component 1`] = `
<button
  className="code-ui-button code-ui-button-warn"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an disabled of dashed 1`] = `
<button
  className="code-ui-button code-ui-button-dashed code-ui-button-dashed-disabled"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an disabled of default 1`] = `
<button
  className="code-ui-button code-ui-button-default code-ui-button-disabled"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders an disabled of primary 1`] = `
<button
  className="code-ui-button code-ui-button-primary code-ui-button-disabled"
  onClick={[Function]}
>
  <span />
</button>
`;

exports[`<Button/> Component render renders loading 1`] = `
<button
  className="code-ui-button code-ui-button-primary"
  onClick={[Function]}
>
  <div
    className="code-ui-button-loading"
  >
    <svg
      className="code-ui-svg"
      viewBox="0 0 1024 1024"
    >
      <path
        d="M537.568 64.32a25.6 25.6 0 000 51.136c204.48 0 370.816 166.368 370.816 370.816a25.6 25.6 0 0051.136 0c0-232.672-189.312-421.984-421.984-421.984z"
      />
    </svg>
  </div>
  <span />
</button>
`;
